[SPARK-28133] Adding inverse hyperbolic functions in SQL

Tonix517 opened the pull request:
## What changes were proposed in this pull request?

Inverse hyperbolic functions (asinh, acosh, atanh) were missing from SQL. Now we add support to those functions.

## How was this patch tested?

Added new tests.
./build/sbt "testOnly org.apache.spark.sql.catalyst.expressions.MathExpressionsSuite
./build/sbt "testOnly org.apache.spark.sql.MathFunctionsSuite

[In spark-shell] 
spark.sql("select asinh(xx)").show
spark.sql("select acosh(xx)").show
spark.sql("select atanh(xx)").show

Please review https://spark.apache.org/contributing.html before opening a pull request.

@wangyum @mgaido91

Commit:
[SPARK-28133] Adding inverse hyperbolic functions in SQL

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

HyukjinKwon commented on python/pyspark/sql/functions.py:
```diff
@@ -187,6 +187,15 @@ def _():
 }
 
+_functions_3_0_over_column = {
+    'asinh': """:param col: hyperbolic angle
```
Can we avoid to add them per https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/functions.scala#L42-L56?

Commit:
[SPARK-28133] Adding inverse hyperbolic functions in SQL

WeichenXu123 committed:
[SPARK-28185][PYTHON][SQL] Closes the generator when Python UDFs stop early

## What changes were proposed in this pull request?

 Closes the generator when Python UDFs stop early.

### Manually verification on pandas iterator UDF and mapPartitions

```python
from pyspark.sql import SparkSession
from pyspark.sql.functions import pandas_udf, PandasUDFType
from pyspark.sql.functions import col, udf
from pyspark.taskcontext import TaskContext
import time
import os

spark.conf.set('spark.sql.execution.arrow.maxRecordsPerBatch', '1')
spark.conf.set('spark.sql.pandas.udf.buffer.size', '4')

pandas_udf("int", PandasUDFType.SCALAR_ITER)
def fi1(it):
    try:
        for batch in it:
            yield batch + 100
            time.sleep(1.0)
    except BaseException as be:
        print("Debug: exception raised: " + str(type(be)))
        raise be
    finally:
        open("/tmp/000001.tmp", "a").close()

df1 = spark.range(10).select(col('id').alias('a')).repartition(1)

# will see log Debug: exception raised: <class 'GeneratorExit'>
# and file "/tmp/000001.tmp" generated.
df1.select(col('a'), fi1('a')).limit(2).collect()

def mapper(it):
    try:
        for batch in it:
                yield batch
    except BaseException as be:
        print("Debug: exception raised: " + str(type(be)))
        raise be
    finally:
        open("/tmp/000002.tmp", "a").close()

df2 = spark.range(10000000).repartition(1)

# will see log Debug: exception raised: <class 'GeneratorExit'>
# and file "/tmp/000002.tmp" generated.
df2.rdd.mapPartitions(mapper).take(2)

```

## How was this patch tested?

Unit test added.

Please review https://spark.apache.org/contributing.html before opening a pull request.

Closes #24986 from WeichenXu123/pandas_iter_udf_limit.

Authored-by: WeichenXu <weichen.xu@databricks.com>
Signed-off-by: HyukjinKwon <gurwls223@apache.org>

mgaido91 commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/mathExpressions.scala:
```diff
@@ -557,6 +578,27 @@ case class Sin(child: Expression) extends UnaryMathExpression(math.sin, "SIN")
+    Examples:
+      > SELECT _FUNC_(0);
+       0.0
+  """)
```
please add `since`

mgaido91 commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/mathExpressions.scala:
```diff
@@ -557,6 +578,27 @@ case class Sin(child: Expression) extends UnaryMathExpression(math.sin, "SIN")
+  extends UnaryMathExpression((x: Double) => math.log(x + math.sqrt(x * x + 1.0)), "ASINH") {
+  override def doGenCode(ctx: CodegenContext, ev: ExprCode): ExprCode = {
+    defineCodeGen(ctx, ev, c =>
+      s"${ev.value} = java.lang.Math.log($c + java.lang.Math.sqrt($c * $c + 1.0));")
```
what if the input is negative?

Tonix517 replied:
It will return NaN, it is the same behavior as for existing sinh implementation.

mgaido91 replied:
mmh..what do other DBs do in this case? I checked the implementation of `FastMath` and it handles negatives...

Tonix517 replied:
thanks for reminding me of FastMath lib. Maybe we can simply use that lib here, without this hand-crafted calculation?

Tonix517 replied:
Also I just tried the code with several negative values, i got correct return values.

mgaido91 replied:
> Also I just tried the code with several negative values, i got correct return values.

Sorry what did you try?

Anyway, if we don't have a dependency on it, I am not sure it is a good idea to add a dependency for only a couple of functions...

Tonix517 replied:
took a closer look at asinh\atanh implementation in FastMath. the negative handling inside is to apply calculation optimization on abs(double a) and then apply proper sign at the end of the function to return, which should produce the same value as the original formula i'm using here (https://en.wikipedia.org/wiki/Inverse_hyperbolic_functions#Definitions_in_terms_of_logarithms). But FastMath runs faster than the original functions though.

mgaido91 commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/mathExpressions.scala:
```diff
@@ -287,6 +287,27 @@ case class Cos(child: Expression) extends UnaryMathExpression(math.cos, "COS")
+    Examples:
+      > SELECT _FUNC_(1);
+       0.0
+  """)
```
ditto

mgaido91 commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/mathExpressions.scala:
```diff
@@ -617,6 +659,27 @@ case class Cot(child: Expression)
+       0.0
+  """)
+case class Atanh(child: Expression)
+  extends UnaryMathExpression((x: Double) => 0.5 * math.log((1.0 + x) / (1.0 - x)), "ATANH") {
```
again, what if the input is negative?

Tonix517 replied:
same as above.

beliefer committed:
[SPARK-28077][SQL] Support ANSI SQL OVERLAY function.

## What changes were proposed in this pull request?

The `OVERLAY` function is a `ANSI` `SQL`.
For example:
```
SELECT OVERLAY('abcdef' PLACING '45' FROM 4);

SELECT OVERLAY('yabadoo' PLACING 'daba' FROM 5);

SELECT OVERLAY('yabadoo' PLACING 'daba' FROM 5 FOR 0);

SELECT OVERLAY('babosa' PLACING 'ubb' FROM 2 FOR 4);
```
The results of the above four `SQL` are:
```
abc45f
yabadaba
yabadabadoo
bubba
```

Note: If the input string is null, then the result is null too.

There are some mainstream database support the syntax.
**PostgreSQL:**
https://www.postgresql.org/docs/11/functions-string.html

**Vertica:** https://www.vertica.com/docs/9.2.x/HTML/Content/Authoring/SQLReferenceManual/Functions/String/OVERLAY.htm?zoom_highlight=overlay

**Oracle:**
https://docs.oracle.com/en/database/oracle/oracle-database/19/arpls/UTL_RAW.html#GUID-342E37E7-FE43-4CE1-A0E9-7DAABD000369

**DB2:**
https://www.ibm.com/support/knowledgecenter/SSGMCP_5.3.0/com.ibm.cics.rexx.doc/rexx/overlay.html

There are some show of the PR on my production environment.
```
spark-sql> SELECT OVERLAY('abcdef' PLACING '45' FROM 4);
abc45f
Time taken: 6.385 seconds, Fetched 1 row(s)
spark-sql> SELECT OVERLAY('yabadoo' PLACING 'daba' FROM 5);
yabadaba
Time taken: 0.191 seconds, Fetched 1 row(s)
spark-sql> SELECT OVERLAY('yabadoo' PLACING 'daba' FROM 5 FOR 0);
yabadabadoo
Time taken: 0.186 seconds, Fetched 1 row(s)
spark-sql> SELECT OVERLAY('babosa' PLACING 'ubb' FROM 2 FOR 4);
bubba
Time taken: 0.151 seconds, Fetched 1 row(s)
spark-sql> SELECT OVERLAY(null PLACING '45' FROM 4);
NULL
Time taken: 0.22 seconds, Fetched 1 row(s)
spark-sql> SELECT OVERLAY(null PLACING 'daba' FROM 5);
NULL
Time taken: 0.157 seconds, Fetched 1 row(s)
spark-sql> SELECT OVERLAY(null PLACING 'daba' FROM 5 FOR 0);
NULL
Time taken: 0.254 seconds, Fetched 1 row(s)
spark-sql> SELECT OVERLAY(null PLACING 'ubb' FROM 2 FOR 4);
NULL
Time taken: 0.159 seconds, Fetched 1 row(s)
```

## How was this patch tested?

Exists UT and new UT.

Closes #24918 from beliefer/ansi-sql-overlay.

Lead-authored-by: gengjiaan <gengjiaan@360.cn>
Co-authored-by: Jiaan Geng <beliefer@163.com>
Signed-off-by: Takuya UESHIN <ueshin@databricks.com>

onursatici committed:
[SPARK-28145][K8S] safe runnable in polling executor source

## What changes were proposed in this pull request?

Add error handling to `ExecutorPodsPollingSnapshotSource`

Closes #24952 from onursatici/os/polling-source.

Authored-by: Onur Satici <onursatici@gmail.com>
Signed-off-by: Sean Owen <sean.owen@databricks.com>

revans2 committed:
[SPARK-27945][SQL] Minimal changes to support columnar processing

## What changes were proposed in this pull request?

This is the first part of [SPARK-27396](https://issues.apache.org/jira/browse/SPARK-27396).  This is the minimum set of changes necessary to support a pluggable back end for columnar processing.  Follow on JIRAs would cover removing some of the duplication between functionality in this patch and functionality currently covered by things like ColumnarBatchScan.

## How was this patch tested?

I added in a new unit test to cover new code not really covered in other places.

I also did manual testing by implementing two plugins/extensions that take advantage of the new APIs to allow for columnar processing for some simple queries.  One version runs on the [CPU](https://gist.github.com/revans2/c3cad77075c4fa5d9d271308ee2f1b1d).  The other version run on a GPU, but because it has unreleased dependencies I will not include a link to it yet.

The CPU version I would expect to add in as an example with other documentation in a follow on JIRA

This is contributed on behalf of NVIDIA Corporation.

Closes #24795 from revans2/columnar-basic.

Authored-by: Robert (Bobby) Evans <bobby@apache.org>
Signed-off-by: Thomas Graves <tgraves@apache.org>

Commit:
[SPARK-28133] Adding inverse hyperbolic functions in SQL

mengxr committed:
[SPARK-28056][.2][PYTHON][SQL] add docstring/doctest for SCALAR_ITER Pandas UDF

## What changes were proposed in this pull request?

Add docstring/doctest for `SCALAR_ITER` Pandas UDF. I explicitly mentioned that per-partition execution is an implementation detail, not guaranteed. I will submit another PR to add the same to user guide, just to keep this PR minimal.

I didn't add "doctest: +SKIP" in the first commit so it is easy to test locally.

cc: HyukjinKwon gatorsmile icexelloss BryanCutler WeichenXu123

![Screen Shot 2019-06-28 at 9 52 41 AM](https://user-images.githubusercontent.com/829644/60358349-b0aa5400-998a-11e9-9ebf-8481dfd555b5.png)
![Screen Shot 2019-06-28 at 9 53 19 AM](https://user-images.githubusercontent.com/829644/60358355-b1db8100-998a-11e9-8f6f-00a11bdbdc4d.png)

## How was this patch tested?

doctest

Closes #25005 from mengxr/SPARK-28056.2.

Authored-by: Xiangrui Meng <meng@databricks.com>
Signed-off-by: gatorsmile <gatorsmile@gmail.com>

HyukjinKwon committed:
[SPARK-28204][SQL][TESTS] Make separate two test cases for column pruning in binary files

## What changes were proposed in this pull request?

SPARK-27534 missed to address my own comments at https://github.com/WeichenXu123/spark/pull/8
It's better to push this in since the codes are already cleaned up.

## How was this patch tested?

Unittests fixed

Closes #25003 from HyukjinKwon/SPARK-27534.

Authored-by: HyukjinKwon <gurwls223@apache.org>
Signed-off-by: HyukjinKwon <gurwls223@apache.org>

WangGuangxin committed:
[SPARK-11412][SQL] Support merge schema for ORC

## What changes were proposed in this pull request?

Currently, ORC's `inferSchema` is implemented as randomly choosing one ORC file and reading its schema.

This PR follows the behavior of Parquet, it implements merge schemas logic by reading all ORC files in parallel through a spark job.

Users can enable merge schema by `spark.read.orc("xxx").option("mergeSchema", "true")` or by setting `spark.sql.orc.mergeSchema` to `true`, the prior one has higher priority.

## How was this patch tested?
tested by UT OrcUtilsSuite.scala

Closes #24043 from WangGuangxin/SPARK-11412.

Lead-authored-by: wangguangxin.cn <wangguangxin.cn@gmail.com>
Co-authored-by: wangguangxin.cn <wangguangxin.cn@bytedance.com>
Signed-off-by: gatorsmile <gatorsmile@gmail.com>

srowen reviewed (commented):
These are very niche functions. Does any other DB support it?

wangyum committed:
[SPARK-28184][SQL][TEST] Avoid creating new sessions in SparkMetadataOperationSuite

## What changes were proposed in this pull request?

To make the #24972 change smaller. This pr improves `SparkMetadataOperationSuite` to avoid creating new sessions when getSchemas/getTables/getColumns.

## How was this patch tested?

N/A

Closes #24985 from wangyum/SPARK-28184.

Authored-by: Yuming Wang <yumwang@ebay.com>
Signed-off-by: gatorsmile <gatorsmile@gmail.com>

wangyum committed:
[SPARK-28196][SQL] Add a new `listTables` and `listLocalTempViews` APIs for SessionCatalog

## What changes were proposed in this pull request?

This pr add two API for [SessionCatalog](https://github.com/apache/spark/blob/df4cb471c9712a2fe496664028d9303caebd8777/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala):
```scala
def listTables(db: String, pattern: String, includeLocalTempViews: Boolean): Seq[TableIdentifier]

def listLocalTempViews(pattern: String): Seq[TableIdentifier]
```
Because in some cases `listTables` does not need local temporary view and sometimes only need list local temporary view.

## How was this patch tested?

unit tests

Closes #24995 from wangyum/SPARK-28196.

Authored-by: Yuming Wang <yumwang@ebay.com>
Signed-off-by: gatorsmile <gatorsmile@gmail.com>

Tonix517 commented:
Copied the link from the JIRA ticket:
https://www.postgresql.org/docs/12/functions-math.html#FUNCTIONS-MATH-HYP-TABLE

Just tried Hive and these functions were not supported indeed.

> These are very niche functions. Does any other DB support it?

LantaoJin committed:
[SPARK-28160][CORE] Fix a bug that callback function may hang when unchecked exception missed

## What changes were proposed in this pull request?

This is very like #23590 .

`ByteBuffer.allocate` may throw `OutOfMemoryError` when the response is large but no enough memory is available. However, when this happens, `TransportClient.sendRpcSync` will just hang forever if the timeout set to unlimited.

This PR catches `Throwable` and uses the error to complete `SettableFuture`.

## How was this patch tested?

I tested in my IDE by setting the value of size to -1 to verify the result. Without this patch, it won't be finished until timeout (May hang forever if timeout set to MAX_INT), or the expected `IllegalArgumentException` will be caught.
```java
Override
      public void onSuccess(ByteBuffer response) {
        try {
          int size = response.remaining();
          ByteBuffer copy = ByteBuffer.allocate(size); // set size to -1 in runtime when debug
          copy.put(response);
          // flip "copy" to make it readable
          copy.flip();
          result.set(copy);
        } catch (Throwable t) {
          result.setException(t);
        }
      }
```

Closes #24964 from LantaoJin/SPARK-28160.

Lead-authored-by: LantaoJin <jinlantao@gmail.com>
Co-authored-by: lajin <lajin@ebay.com>
Signed-off-by: Sean Owen <sean.owen@databricks.com>

mgaido91 committed:
[SPARK-28170][ML][PYTHON] Uniform Vectors and Matrix documentation

## What changes were proposed in this pull request?

The documentation in `linalg.py` is not consistent. This PR uniforms the documentation.

## How was this patch tested?

NA

Closes #25011 from mgaido91/SPARK-28170.

Authored-by: Marco Gaido <marcogaido91@gmail.com>
Signed-off-by: HyukjinKwon <gurwls223@apache.org>

HyukjinKwon commented on sql/core/src/main/scala/org/apache/spark/sql/functions.scala:
```diff
@@ -2023,11 +2041,29 @@ object functions {
-   * @since 1.4.0
+   * @since 3.0.0
    */
+  def asinh(columnName: String): Column = asinh(Column(columnName))
```
Shall we remove those APIs in Scala and R sides?

srowen replied:
They don't exist already, right?

Tonix517 replied:
Hi @HyukjinKwon, do you mean we remove then from R for the same reason we removed them from Python, and just keep them in Scala?

HyukjinKwon replied:
Sorry I had to clarify it. Yes, let's add it only into SQL side and remove them in Scala, Python and R per https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/functions.scala#L42-L56

These are only added into other language APIs when they are commonly used. I think we don't have to add them in the first place in general.

Tonix517 replied:
thanks Hyukjin. just removed them from language APIs. now only available in SQL

HyukjinKwon commented:
Let me leave this triggered. cc @wangyum and @gatorsmile it looks like the same feature party of PostgreSQL

HyukjinKwon commented:
ok to test

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/12295/
Test PASSed.

SparkQA commented:
**[Test build #107119 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/107119/testReport)** for PR 24994 at commit [`8dfaabc`](https://github.com/apache/spark/commit/8dfaabcebc53d3c3b0501a825e4e85a858d7bb76).

SparkQA commented:
**[Test build #107119 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/107119/testReport)** for PR 24994 at commit [`8dfaabc`](https://github.com/apache/spark/commit/8dfaabcebc53d3c3b0501a825e4e85a858d7bb76).
 * This patch **fails R style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/107119/
Test FAILed.

Commit:
[SPARK-28133] Adding inverse hyperbolic functions in SQL

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/12322/
Test PASSed.

SparkQA commented:
**[Test build #107149 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/107149/testReport)** for PR 24994 at commit [`3b07ee3`](https://github.com/apache/spark/commit/3b07ee33322de6d15d80bc65e32413fc5697880c).

HyukjinKwon commented:
Looks something when wrong during rebase. Can you resolve it or open a new PR?

SparkQA commented:
**[Test build #107149 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/107149/testReport)** for PR 24994 at commit [`3b07ee3`](https://github.com/apache/spark/commit/3b07ee33322de6d15d80bc65e32413fc5697880c).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/107149/
Test FAILed.
